### PR TITLE
Implement the first half of batch inference (starting the job)

### PR DESCRIPTION
**Pull request by virajmehta**

### CODE REVIEW

File: gateway/tests/e2e/providers/batch.rs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#![allow(clippy::print_stdout)]
+
+#[macro_export]
+macro_rules! generate_post_batch_inference_tests {
+    ($func:ident) => {
+        todo!()
+    };
+}
```
Comment (GabrielBianconi) [inline gateway/tests/e2e/providers/batch.rs:8]: ?

File: gateway/src/variant/dicl.rs [hunk 2]
```diff
@@ -256,6 +260,18 @@ impl Variant for DiclConfig {
...
+        &'a self,
+        _input: &[Input],
+        _models: &'a InferenceModels<'a>,
+        _function: &'a FunctionConfig,
+        _inference_configs: &'a [InferenceConfig<'a, 'a>],
+        _clients: &'a InferenceClients<'a>,
+        _inference_params: Vec<InferenceParams>,
+    ) -> Result<BatchModelInferenceWithMetadata<'a>, Error> {
+        Err(ErrorDetails::UnsupportedVariantForBatchInference { variant_name: None }.into())
+    }
```
Comment (GabrielBianconi) [inline gateway/src/variant/dicl.rs:274]: Create a TODO/issue for this. This is doable in batch format.

File: gateway/src/variant/mod.rs [hunk 3]
```diff
@@ -255,6 +328,36 @@ impl Variant for VariantConfig {
         }
     }
 
+    // TODO(Viraj): deal with this
```
Comment (GabrielBianconi) [inline gateway/src/variant/mod.rs:331]: ^

File: gateway/src/variant/dicl.rs (hunk 2 above)
Comment (virajmehta) [reply gateway/src/variant/dicl.rs:279]: How?

Comment (virajmehta) [reply gateway/src/variant/dicl.rs:279]: presumably it's not a good solution to embed each input in the batch synchronously?

are we gonna do multistep polling? I think that's not something we should kill ourselves over

File: gateway/src/clickhouse_migration_manager/migrations/migration_0006.rs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        Ok(false)
+    }
+
+    async fn apply(&self) -> Result<(), Error> {
+        // Create the `BatchModelInference` table
+        let query = r#"
+            CREATE TABLE IF NOT EXISTS BatchModelInference
+            (
+                id UUID,
+                batch_id UUID,
```
[the remark this answers is not in the document]
Comment (virajmehta) [reply gateway/src/clickhouse_migration_manager/migrations/migration_0006.rs:79]: changed to inference_id

File: gateway/src/endpoints/batch_inference.rs [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+    }
+
+    // Retrieve or generate the episode IDs and validate them (in the impl)
+    let episode_ids: BatchEpisodeIds =
+        BatchEpisodeIdsWithSize(params.episode_ids, num_inferences).try_into()?;
+
+    // Increment the request count
+    counter!(
+        "request_count",
+        "endpoint" => "post_batch_inference",
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/batch_inference.rs:166]: `endpoint` should be the actual endpoint (`batch_inference`)

We can optionally further parametrize with `method` = `POST`. If we do that, we should do that for every endpoint.

File: gateway/src/endpoints/inference.rs [hunk 6]
```diff
@@ -190,19 +193,28 @@ pub async fn inference(
             "function_name" => params.function_name.to_string(),
         )
         .increment(1);
+        counter!(
+            "inference_count",
+            "endpoint" => "post_batch_inference",
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/inference.rs:198]: Wrong

File: gateway/src/endpoints/batch_inference.rs [hunk 7]
```diff
@@ -0,0 +1,886 @@
...
+        "function_name" => params.function_name.to_string(),
+    )
+    .increment(1);
+    counter!(
+        "inference_count",
+        "endpoint" => "post_batch_inference",
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/batch_inference.rs:172]: Same as above.

File: gateway/src/endpoints/batch_inference.rs [hunk 8]
```diff
@@ -0,0 +1,886 @@
...
+            tags: params.tags,
+        };
+
+        let (batch_id, inference_ids) = write_inference(
+            &clickhouse_connection_info,
+            params.inputs,
+            result,
+            write_metadata,
+            inference_config.clone(),
+            // Spent a while fighting the borrow checker here, gave up
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/batch_inference.rs:257]: TODO

File: gateway/src/endpoints/batch_inference.rs [hunk 9]
```diff
@@ -0,0 +1,886 @@
...
+    episode_ids: Vec<Uuid>,
+}
+
+#[derive(Debug)]
+struct BatchInferenceDatabaseInsertMetadata<'a> {
+    pub function_name: &'a str,
+    pub variant_name: &'a str,
+    pub episode_ids: &'a Vec<Uuid>,
+    pub tags: Option<Vec<Option<HashMap<String, String>>>>,
+    // pub tool_configs: &'a Vec<Option<ToolCallConfig>>,
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/batch_inference.rs:290]: ?

File: gateway/src/endpoints/batch_inference.rs [hunk 10]
```diff
@@ -0,0 +1,886 @@
...
+    pub output_schema: Option<String>,
+    pub model_name: &'a str,
+    pub model_provider_name: &'a str,
+    pub tags: HashMap<String, String>,
+}
+
+#[derive(Debug, Serialize)]
+struct BatchRequestInsert<'a> {
+    batch_id: &'a str,
+    id: String,
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/batch_inference.rs:314]: Why do we need this `id`? Can't we just use `batch_id`?

File: gateway/src/endpoints/batch_inference.rs [hunk 11]
```diff
@@ -0,0 +1,886 @@
...
+        result.input_messages.iter(),
+        result.systems.iter(),
+        inference_config.tool_configs,
+        result.inference_params.iter(),
+        result.output_schemas.iter(),
+        metadata
+            .tags
+            .unwrap_or_default()
+            .into_iter()
+            .chain(repeat(None)),
```
Comment (GabrielBianconi) [inline gateway/src/endpoints/batch_inference.rs:385]: Can you clean this up with a struct?

File: gateway/src/inference/providers/openai.rs [hunk 12]
```diff
@@ -1122,6 +1278,56 @@ impl<'a> TryFrom<OpenAIEmbeddingResponseWithMetadata<'a>> for EmbeddingProviderR
...
+    // data: Vec<OpenAIBatchError>,
+}
+
+#[derive(Debug, Deserialize)]
+struct OpenAIBatchError {
+    // code: String,
+    // message: String,
+    // param: Option<String>,
+    // line: Option<i32>,
+}
```
Comment (GabrielBianconi) [inline gateway/src/inference/providers/openai.rs:1322]: Is stuff like this used anywhere?

File: gateway/src/main.rs [hunk 13]
```diff
@@ -38,6 +38,10 @@ async fn main() {
 
     let router = Router::new()
         .route("/inference", post(endpoints::inference::inference_handler))
+        .route(
+            "/batch_inference",
+            post(endpoints::batch_inference::prepare_batch_inference_handler),
```
Comment (GabrielBianconi) [inline gateway/src/main.rs:43]: You mix `start_batch_inference` `prepare_batch_inference` `post_batch_inference` throughout.

File: gateway/src/variant/dicl.rs (hunk 2 above)
Comment (GabrielBianconi) [reply gateway/src/variant/dicl.rs:279]: It's a reasonable starter solution. Embedding is often much much cheaper. But just TODO it.

**Review by GabrielBianconi — CHANGES_REQUESTED** [review]
Partial review.

Requires a lot more documentation, hard to follow.

File: gateway/src/clickhouse_migration_manager/migrations/migration_0006.rs [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+                    }
+                    .into())
+                }
+                Ok(response) => {
+                    if response.trim() != "1" {
+                        return Ok(true);
+                    }
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (virajmehta) [reply gateway/src/clickhouse_migration_manager/migrations/migration_0006.rs:65]: opened issue

File: gateway/src/inference/providers/openai.rs (hunk 12 above)
Comment (virajmehta) [reply gateway/src/inference/providers/openai.rs:1322]: this was for my reference for the next PR -- not used yet

File: gateway/src/endpoints/batch_inference.rs (hunk 10 above)
Comment (virajmehta) [reply gateway/src/endpoints/batch_inference.rs:314]: this is gonna get appended to with more writes as we poll so yes

**Comment by virajmehta** [comment]
> Partial review.
> 
> Requires a lot more documentation, hard to follow.

I'll add proper module-level documentation during the next PR once the story is finalized. But I'll add some now.

File: gateway/src/endpoints/batch_inference.rs (hunk 11 above)
Comment (virajmehta) [reply gateway/src/endpoints/batch_inference.rs]: done, still not that neat but at least we collect then serialize

**Review by GabrielBianconi — APPROVED** [review]
